lib/bundle/locker.rb

```diff
@@ -20,25 +22,58 @@ def lockfile(global: false, file: nil)
       lock_file_path
     end
 
-    def write_lockfile?(global: false, file: nil, no_lock: false)
+    def write_lockfile?(path: nil, no_lock: false)
       return false if no_lock
       return false if ENV["HOMEBREW_BUNDLE_NO_LOCK"]
 
       # handle the /dev/stdin and /dev/stdout cases
-      return false if lockfile(global: global, file: file).parent.to_s == "/dev"
+      return false if path.nil? || path.parent.to_s == "/dev"
+
+      true
+    end
 
+    def write_lockfile!(lockfile, lock_data)
+      json = JSON.pretty_generate(lock_data)
+      begin
+        lockfile.unlink if lockfile.exist?
+        lockfile.write("#{json}\n")
+      rescue Errno::EPERM, Errno::EACCES, Errno::ENOTEMPTY
+        opoo "Could not write to #{lockfile}!"
+        return false
+      end
       true
     end
 
+    def read_lockfile_if_exists(path)
+      return JSON.parse(path.read) if path.exist?
+
+      {}
+    end
+
     def lock(entries, global: false, file: nil, no_lock: false)
-      return false unless write_lockfile?(global: global, file: file, no_lock: no_lock)
+      lockfile = resolve_lockfile_path(global: global, brewfile: file)
 
-      lockfile = lockfile(global: global, file: file)
+      return false unless write_lockfile?(path: lockfile, no_lock: no_lock)
 
-      lock = JSON.parse(lockfile.read) if lockfile.exist?
-      lock ||= {}
+      lock_data = build_lock_data(entries, base: read_lockfile_if_exists(lockfile))
+
+      write_lockfile!(lockfile, lock_data)
+    end
+
+    def build_lock_data(entries, base: {})
+      lock ||= base
+      lock = update_lock_metadata(lock: lock)
+      lock = add_entries_data(entries, lock: lock)
+      add_system_data(lock: lock)
+    end
+
+    def update_lock_metadata(lock: {})
+      lock["version"] = LOCKFILE_FORMAT_VERSION
+      lock
+    end
+
+    def add_entries_data(entries, lock: {})
```

Comment:
On brevity, I fought the urge to refactor this, and then wound up running out of time, so that was a worthwhile effort dodged.